# Lab | Data Structuring and Combining Data

## Challenge 1: Combining & Cleaning Data

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the links below.

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Note that you'll need to clean and format the new data.

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

In [81]:
import pandas as pd

url = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv'
url_2 = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv'
url_3 = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv'

# Read the CSV file into a DataFrame
customer_data_df = pd.read_csv(url)
display(customer_data_df)

customer_data_df_2 = pd.read_csv(url_2)
display(customer_data_df_2)

customer_data_df_3 = pd.read_csv(url_3)
display(customer_data_df_3)


Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     Number of Open Complaints     Policy Type  Vehicle Class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

Customer          ST GENDER Education Customer Lifetime Value  Income  \
0    GS98873     Arizona      F  Bachelor              323912.47%   16061   
1    CW49887  California      F    Master              462680.11%   79487   
2    MY31220  California      F   College              899704.02%   54230   
3    UH35128      Oregon      F   College             2580706.30%   71210   
4    WH52799     Arizona      F   College              380812.21%   94903   
..       ...         ...    ...       ...                     ...     ...   
991  HV85198     Arizona      M    Master              847141.75%   63513   
992  BS91566     Arizona      F   College              543121.91%   58161   
993  IL40123      Nevada      F   College              568964.41%   83640   
994  MY32149  California      F    Master              368672.38%       0   
995  SA91515  California      M  Bachelor              399258.39%       0   

     Monthly Premium Auto Number of Open Complaints  Total Claim Amount  \
0                      88                    1/0/00          633.600000   
1                     114                    1/0/00          547.200000   
2                     112                    1/0/00          537.600000   
3                     214                    1/1/00         1027.200000   
4                      94                    1/0/00          451.200000   
..                    ...                       ...                 ...   
991                    70                    1/0/00          185.667213   
992                    68                    1/0/00          140.747286   
993                    70                    1/0/00          471.050488   
994                    96                    1/0/00           28.460568   
995                   111                    1/0/00          700.349052   

        Policy Type  Vehicle Class  
0     Personal Auto  Four-Door Car  
1      Special Auto            SUV  
2     Personal Auto   Two-Door Car  
3     Personal Auto     Luxury Car  
4    Corporate Auto   Two-Door Car  
..              ...            ...  
991   Personal Auto  Four-Door Car  
992  Corporate Auto  Four-Door Car  
993  Corporate Auto   Two-Door Car  
994   Personal Auto   Two-Door Car  
995   Personal Auto            SUV  

[996 rows x 11 columns]

Customer       State  Customer Lifetime Value             Education  \
0     SA25987  Washington              3479.137523  High School or Below   
1     TB86706     Arizona              2502.637401                Master   
2     ZL73902      Nevada              3265.156348              Bachelor   
3     KX23516  California              4455.843406  High School or Below   
4     FN77294  California              7704.958480  High School or Below   
...       ...         ...                      ...                   ...   
7065  LA72316  California             23405.987980              Bachelor   
7066  PK87824  California              3096.511217               College   
7067  TD14365  California              8163.890428              Bachelor   
7068  UP19263  California              7524.442436               College   
7069  Y167826  California              2611.836866               College   

     Gender  Income  Monthly Premium Auto  Number of Open Complaints  \
0         M       0                   104                          0   
1         M       0                    66                          0   
2         F   25820                    82                          0   
3         F       0                   121                          0   
4         M   30366                   101                          2   
...     ...     ...                   ...                        ...   
7065      M   71941                    73                          0   
7066      F   21604                    79                          0   
7067      M       0                    85                          3   
7068      M   21941                    96                          0   
7069      M       0                    77                          0   

         Policy Type  Total Claim Amount  Vehicle Class  
0      Personal Auto          499.200000   Two-Door Car  
1      Personal Auto            3.468912   Two-Door Car  
2      Personal Auto          393.600000  Four-Door Car  
3      Personal Auto          699.615192            SUV  
4      Personal Auto          484.800000            SUV  
...              ...                 ...            ...  
7065   Personal Auto          198.234764  Four-Door Car  
7066  Corporate Auto          379.200000  Four-Door Car  
7067  Corporate Auto          790.784983  Four-Door Car  
7068   Personal Auto          691.200000  Four-Door Car  
7069  Corporate Auto          369.600000   Two-Door Car  

[7070 rows x 11 columns]

In [82]:
#renaming columns for 3

#renaming columns for 1 & 2

customer_data_df_3.rename(columns = {'Customer': 'customer', 
                         'State' : 'state', 
                         'Gender' : 'gender',
                         'Education' : 'education',
                         'Customer Lifetime Value' : 'customer_lifetime_value',
                         'Income' : 'income',
                         'Monthly Premium Auto' : 'monthly_premium_auto',
                         'Number of Open Complaints' : 'number_of_open_complaints',
                         'Policy Type' : 'policy_type',
                         'Vehicle Class' : 'vehicle_class',
                         'Total Claim Amount' : 'total_claim_amount'}, inplace=True)


In [83]:
#merging df 1 and df 2
df_1_2 = pd.concat([customer_data_df, customer_data_df_2], axis=0)
display(df_1_2)

df_1_2['Vehicle Class'].value_counts()

Customer          ST GENDER             Education Customer Lifetime Value  \
0    RB50392  Washington    NaN                Master                     NaN   
1    QZ44356     Arizona      F              Bachelor              697953.59%   
2    AI49188      Nevada      F              Bachelor             1288743.17%   
3    WW63253  California      M              Bachelor              764586.18%   
4    GA49547  Washington      M  High School or Below              536307.65%   
..       ...         ...    ...                   ...                     ...   
991  HV85198     Arizona      M                Master              847141.75%   
992  BS91566     Arizona      F               College              543121.91%   
993  IL40123      Nevada      F               College              568964.41%   
994  MY32149  California      F                Master              368672.38%   
995  SA91515  California      M              Bachelor              399258.39%   

      Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0        0.0                1000.0                    1/0/00   Personal Auto   
1        0.0                  94.0                    1/0/00   Personal Auto   
2    48767.0                 108.0                    1/0/00   Personal Auto   
3        0.0                 106.0                    1/0/00  Corporate Auto   
4    36357.0                  68.0                    1/0/00   Personal Auto   
..       ...                   ...                       ...             ...   
991  63513.0                  70.0                    1/0/00   Personal Auto   
992  58161.0                  68.0                    1/0/00  Corporate Auto   
993  83640.0                  70.0                    1/0/00  Corporate Auto   
994      0.0                  96.0                    1/0/00   Personal Auto   
995      0.0                 111.0                    1/0/00   Personal Auto   

     Vehicle Class  Total Claim Amount  
0    Four-Door Car            2.704934  
1    Four-Door Car         1131.464935  
2     Two-Door Car          566.472247  
3              SUV          529.881344  
4    Four-Door Car           17.269323  
..             ...                 ...  
991  Four-Door Car          185.667213  
992  Four-Door Car          140.747286  
993   Two-Door Car          471.050488  
994   Two-Door Car           28.460568  
995            SUV          700.349052  

[5004 rows x 11 columns]

Vehicle Class
Four-Door Car    1095
Two-Door Car      416
SUV               373
Sports Car        112
Luxury SUV         41
Luxury Car         30
Name: count, dtype: int64

In [84]:
#renaming columns for 1 & 2

df_1_2.rename(columns = {'Customer': 'customer', 
                         'ST' : 'state', 
                         'GENDER' : 'gender',
                         'Education' : 'education',
                         'Customer Lifetime Value' : 'customer_lifetime_value',
                         'Income' : 'income',
                         'Monthly Premium Auto' : 'monthly_premium_auto',
                         'Number of Open Complaints' : 'number_of_open_complaints',
                         'Policy Type' : 'policy_type',
                         'Vehicle Class' : 'vehicle_class',
                         'Total Claim Amount' : 'total_claim_amount'}, inplace=True)
df_1_2


customer       state gender             education customer_lifetime_value  \
0    RB50392  Washington    NaN                Master                     NaN   
1    QZ44356     Arizona      F              Bachelor              697953.59%   
2    AI49188      Nevada      F              Bachelor             1288743.17%   
3    WW63253  California      M              Bachelor              764586.18%   
4    GA49547  Washington      M  High School or Below              536307.65%   
..       ...         ...    ...                   ...                     ...   
991  HV85198     Arizona      M                Master              847141.75%   
992  BS91566     Arizona      F               College              543121.91%   
993  IL40123      Nevada      F               College              568964.41%   
994  MY32149  California      F                Master              368672.38%   
995  SA91515  California      M              Bachelor              399258.39%   

      income  monthly_premium_auto number_of_open_complaints     policy_type  \
0        0.0                1000.0                    1/0/00   Personal Auto   
1        0.0                  94.0                    1/0/00   Personal Auto   
2    48767.0                 108.0                    1/0/00   Personal Auto   
3        0.0                 106.0                    1/0/00  Corporate Auto   
4    36357.0                  68.0                    1/0/00   Personal Auto   
..       ...                   ...                       ...             ...   
991  63513.0                  70.0                    1/0/00   Personal Auto   
992  58161.0                  68.0                    1/0/00  Corporate Auto   
993  83640.0                  70.0                    1/0/00  Corporate Auto   
994      0.0                  96.0                    1/0/00   Personal Auto   
995      0.0                 111.0                    1/0/00   Personal Auto   

     vehicle_class  total_claim_amount  
0    Four-Door Car            2.704934  
1    Four-Door Car         1131.464935  
2     Two-Door Car          566.472247  
3              SUV          529.881344  
4    Four-Door Car           17.269323  
..             ...                 ...  
991  Four-Door Car          185.667213  
992  Four-Door Car          140.747286  
993   Two-Door Car          471.050488  
994   Two-Door Car           28.460568  
995            SUV          700.349052  

[5004 rows x 11 columns]

In [85]:
# assigning only M and F for gender

gender_mapping = {'male': 'M', 'female': 'F', 'Femal' : 'F', 'Male' : 'M', 'Female' : 'F'}
df_1_2['gender'] = df_1_2['gender'].replace(gender_mapping)
df_1_2['gender'].unique()

array([nan, 'F', 'M'], dtype=object)

In [86]:
# replacing abbreviations of states 

replacing_abbreviations = {'AZ': 'Arizona', 'Cali': 'California', 'WA' : 'Washington'}
df_1_2['state'] = df_1_2['state'].replace(replacing_abbreviations)
display(df_1_2['state'])


0      Washington
1         Arizona
2          Nevada
3      California
4      Washington
          ...    
991       Arizona
992       Arizona
993        Nevada
994    California
995    California
Name: state, Length: 5004, dtype: object

In [87]:
#*In education, "Bachelors" could be replaced by "Bachelor"*

rename_education = {'Bachelors': 'Bachelor'}
df_1_2['education'] = df_1_2['education'].replace(rename_education)
display(df_1_2['education'])

0                    Master
1                  Bachelor
2                  Bachelor
3                  Bachelor
4      High School or Below
               ...         
991                  Master
992                 College
993                 College
994                  Master
995                Bachelor
Name: education, Length: 5004, dtype: object

In [88]:
#*In Customer Lifetime Value, delete the `%` character*

df_1_2['customer_lifetime_value'] = df_1_2['customer_lifetime_value'].str.replace('%', '')
display(df_1_2['customer_lifetime_value'])

0             NaN
1       697953.59
2      1288743.17
3       764586.18
4       536307.65
          ...    
991     847141.75
992     543121.91
993     568964.41
994     368672.38
995     399258.39
Name: customer_lifetime_value, Length: 5004, dtype: object

In [89]:
#Number of open complaints formating 

df_1_2['number_of_open_complaints'] = df_1_2['number_of_open_complaints'].fillna(0).astype(int)


ValueError: invalid literal for int() with base 10: '1/0/00'

In [ ]:
# merging df 1, 2 & 3 using outer

df_1_2_3 = pd.concat([df_1_2, customer_data_df_3], axis=0, join="outer")
df_1_2_3

customer       state gender             education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
7065  LA72316  California      M              Bachelor   
7066  PK87824  California      F               College   
7067  TD14365  California      M              Bachelor   
7068  UP19263  California      M               College   
7069  Y167826  California      M               College   

     customer_lifetime_value   income  monthly_premium_auto  \
0                        NaN      0.0                1000.0   
1                  697953.59      0.0                  94.0   
2                 1288743.17  48767.0                 108.0   
3                  764586.18      0.0                 106.0   
4                  536307.65  36357.0                  68.0   
...                      ...      ...                   ...   
7065             23405.98798  71941.0                  73.0   
7066             3096.511217  21604.0                  79.0   
7067             8163.890428      0.0                  85.0   
7068             7524.442436  21941.0                  96.0   
7069             2611.836866      0.0                  77.0   

      number_of_open_complaints     policy_type  vehicle_class  \
0                             0   Personal Auto  Four-Door Car   
1                             0   Personal Auto  Four-Door Car   
2                             0   Personal Auto   Two-Door Car   
3                             0  Corporate Auto            SUV   
4                             0   Personal Auto  Four-Door Car   
...                         ...             ...            ...   
7065                          0   Personal Auto  Four-Door Car   
7066                          0  Corporate Auto  Four-Door Car   
7067                          3  Corporate Auto  Four-Door Car   
7068                          0   Personal Auto  Four-Door Car   
7069                          0  Corporate Auto   Two-Door Car   

      total_claim_amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
7065          198.234764  
7066          379.200000  
7067          790.784983  
7068          691.200000  
7069          369.600000  

[12074 rows x 11 columns]

# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

In [ ]:
df_1_2_3


customer       state gender             education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
7065  LA72316  California      M              Bachelor   
7066  PK87824  California      F               College   
7067  TD14365  California      M              Bachelor   
7068  UP19263  California      M               College   
7069  Y167826  California      M               College   

     customer_lifetime_value   income  monthly_premium_auto  \
0                        NaN      0.0                1000.0   
1                  697953.59      0.0                  94.0   
2                 1288743.17  48767.0                 108.0   
3                  764586.18      0.0                 106.0   
4                  536307.65  36357.0                  68.0   
...                      ...      ...                   ...   
7065             23405.98798  71941.0                  73.0   
7066             3096.511217  21604.0                  79.0   
7067             8163.890428      0.0                  85.0   
7068             7524.442436  21941.0                  96.0   
7069             2611.836866      0.0                  77.0   

      number_of_open_complaints     policy_type  vehicle_class  \
0                             0   Personal Auto  Four-Door Car   
1                             0   Personal Auto  Four-Door Car   
2                             0   Personal Auto   Two-Door Car   
3                             0  Corporate Auto            SUV   
4                             0   Personal Auto  Four-Door Car   
...                         ...             ...            ...   
7065                          0   Personal Auto  Four-Door Car   
7066                          0  Corporate Auto  Four-Door Car   
7067                          3  Corporate Auto  Four-Door Car   
7068                          0   Personal Auto  Four-Door Car   
7069                          0  Corporate Auto   Two-Door Car   

      total_claim_amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
7065          198.234764  
7066          379.200000  
7067          790.784983  
7068          691.200000  
7069          369.600000  

[12074 rows x 11 columns]

In [ ]:
url = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv'

marketing_df = pd.read_csv(url)
display(marketing_df)


In [ ]:
# rename employment status

marketing_df.rename(columns = {'employmentstatus': 'employment_status'}, inplace=True)


In [ ]:
marketing_df.info()

In [ ]:
marketing_df.drop('unnamed:_0', axis=1)

customer       state  customer_lifetime_value response  coverage  \
0      DK49336     Arizona              4809.216960       No     Basic   
1      KX64629  California              2228.525238       No     Basic   
2      LZ68649  Washington             14947.917300       No     Basic   
3      XL78013      Oregon             22332.439460      Yes  Extended   
4      QA50777      Oregon              9025.067525       No   Premium   
...        ...         ...                      ...      ...       ...   
10905  FE99816      Nevada             15563.369440       No   Premium   
10906  KX53892      Oregon              5259.444853       No     Basic   
10907  TL39050     Arizona             23893.304100       No  Extended   
10908  WA60547  California             11971.977650       No   Premium   
10909  IV32877  California              6857.519928       No     Basic   

      education effective_to_date employment_status gender  income  ...  \
0       College        2011-02-18          Employed      M   48029  ...   
1       College        2011-01-18        Unemployed      F       0  ...   
2      Bachelor        2011-02-10          Employed      M   22139  ...   
3       College        2011-01-11          Employed      M   49078  ...   
4      Bachelor        2011-01-17     Medical Leave      F   23675  ...   
...         ...               ...               ...    ...     ...  ...   
10905  Bachelor        2011-01-19        Unemployed      F       0  ...   
10906   College        2011-01-06          Employed      F   61146  ...   
10907  Bachelor        2011-02-06          Employed      F   39837  ...   
10908   College        2011-02-13          Employed      F   64195  ...   
10909  Bachelor        2011-01-08        Unemployed      M       0  ...   

      number_of_policies     policy_type        policy  renew_offer_type  \
0                      9  Corporate Auto  Corporate L3            Offer3   
1                      1   Personal Auto   Personal L3            Offer4   
2                      2   Personal Auto   Personal L3            Offer3   
3                      2  Corporate Auto  Corporate L3            Offer2   
4                      7   Personal Auto   Personal L2            Offer1   
...                  ...             ...           ...               ...   
10905                  7   Personal Auto   Personal L1            Offer3   
10906                  6   Personal Auto   Personal L3            Offer2   
10907                  2  Corporate Auto  Corporate L3            Offer1   
10908                  6   Personal Auto   Personal L1            Offer1   
10909                  3   Personal Auto   Personal L1            Offer4   

       sales_channel  total_claim_amount  vehicle_class vehicle_size  \
0              Agent          292.800000  Four-Door Car      Medsize   
1        Call Center          744.924331  Four-Door Car      Medsize   
2        Call Center          480.000000            SUV      Medsize   
3             Branch          484.013411  Four-Door Car      Medsize   
4             Branch          707.925645  Four-Door Car      Medsize   
...              ...                 ...            ...          ...   
10905            Web         1214.400000     Luxury Car      Medsize   
10906         Branch          273.018929  Four-Door Car      Medsize   
10907            Web          381.306996     Luxury SUV      Medsize   
10908         Branch          618.288849            SUV      Medsize   
10909            Web         1021.719397            SUV      Medsize   

      vehicle_type month  
0                A     2  
1                A     1  
2                A     2  
3                A     1  
4                A     1  
...            ...   ...  
10905            A     1  
10906            A     1  
10907            A     2  
10908            A     2  
10909            A     1  

[10910 rows x 26 columns]

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

In [ ]:
# analyze which sales channel brought the most sales 
pivot_marketing_df = marketing_df.pivot_table(values='total_claim_amount', index='sales_channel', aggfunc='sum').round(2)

display(pivot_marketing_df)

total_claim_amount
sales_channel                    
Agent                  1810226.82
Branch                 1301204.00
Call Center             926600.82
Web                     706600.04


### Analysis and Insights:

1. **Highest Total Claims - Agent**:
   - The **Agent** channel has the highest total claim amount of **$1,810,226.82**.
   - This indicates that the Agent channel is handling the most significant volume of claims, suggesting it may have the most substantial interaction with customers or deal with higher-value claims.

2. **Branch Performance**:
   - The **Branch** channel has the second highest total claim amount of **$1,301,204.00**.
   - Branch offices are likely serving a large number of customers, making it a critical channel for the company. This may also indicate that customers prefer in-person service for their claims.

3. **Call Center**:
   - The **Call Center** has a total claim amount of **$926,600.82**.
   - Call centers handle a significant volume of claims, indicating that many customers prefer resolving their claims over the phone, perhaps for convenience or efficiency.

4. **Web Channel**:
   - The **Web** channel has the lowest total claim amount of **$706,600.04**.
   - Although it has the lowest total, it shows that a considerable number of customers are comfortable using online services for their claims. This could represent a growth opportunity, especially as more people become accustomed to digital services.

### Further Insights:

1. **Channel Effectiveness**:
   - Agents are likely highly effective in resolving claims and may have better customer service, leading to a higher volume of processed claims.
   - Branches, while second, might offer a balance between in-person service and volume.
   - Call Centers offer a middle ground, providing convenience without the need for physical presence.
   - Web services, although currently the lowest, could be the most cost-effective and scalable with the right improvements in user experience and trust.

2. **Customer Preferences**:
   - Customers might prefer personal interaction (Agent and Branch) for claims, potentially because of the perceived importance and complexity of the claims process.
   - The lower total for Web claims suggests either a lower adoption rate of the digital channel or lower trust/comfort levels with handling claims online.

3. **Strategic Decisions**:
   - Investing in enhancing the web channel could yield significant benefits, as it is cost-effective and scalable.
   - Training and resource allocation to agents should continue, given their high performance.
   - Branches remain important for those who prefer face-to-face interactions.
   - Call centers might benefit from additional support and technology to handle a growing volume of claims.

### Recommendations:

1. **Enhance Digital Presence**:
   - Improve the web platform to make it more user-friendly and secure to increase its adoption for claim processing.
   
2. **Support High-Performing Channels**:
   - Continue to support and possibly expand the agent network and branch services due to their high claim volumes.

3. **Increase Efficiency**:
   - Optimize call center operations with better training and technology to handle claims more efficiently and increase customer satisfaction.

4. **Marketing and Outreach**:
   - Promote the web channel's ease and efficiency to encourage more customers to use it.
   - Highlight the advantages of each channel to ensure customers are aware of all their options.

By focusing on these areas, the company can improve its overall efficiency, customer satisfaction, and potentially increase the total claim amounts processed across all channels.

2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

In [ ]:
# calculate the average customer lifetime value for gender and education

pivot_marketing_df = marketing_df.pivot_table(values='customer_lifetime_value', columns= 'gender', index='education', aggfunc='mean').round(2)


Analysis and Insights:
Education Level Impact on CLV:

High School or Below: Both females and males have the highest average CLV among all education levels. Females average $8,675.22, slightly higher than males at $8,149.69. This suggests that individuals with a high school education or lower contribute significantly to CLV.

Bachelor's Degree: Females have a slightly higher average CLV ($7,874.27) compared to males ($7,703.60), indicating that individuals with a bachelor's degree also contribute substantially to CLV.

College: Females have an average CLV of $7,748.82, whereas males have $8,052.46. This shows a slight difference in CLV between genders at the college education level.

Doctor: Females average $7,328.51, while males average $7,415.33. This indicates a relatively balanced CLV between genders among those with a doctorate.

Master's Degree: Females have an average CLV of $8,157.05, and males have $8,168.83. This represents a very similar CLV between genders at the master's degree level.

Gender Comparison:

Across all education levels, females generally have comparable or slightly higher average CLV compared to males. This suggests that females may have a slightly higher lifetime value as customers in the insurance dataset.
Educational Attainment Trends:

Customers with higher educational attainment levels (Master's, Doctorate) generally exhibit slightly lower average CLV compared to those with lower educational levels (High School or Below, Bachelor's). This contrasts with some common expectations about the relationship between education and income.
Strategic Implications:

Targeting Strategies: Understanding the different CLV by education and gender can help tailor marketing and retention strategies. For example, focusing on retaining high school-educated customers, who contribute significantly to CLV.

Product Offerings: Adjusting product offerings or service enhancements based on educational attainment levels may attract and retain customers in different education categories.

Customer Segmentation: Using this data, marketers can segment their customer base effectively to maximize CLV through targeted campaigns and personalized communication strategies.

Further Analysis:

It would be beneficial to investigate further into the underlying reasons for the differences in CLV among different education levels and genders. Factors such as customer behaviors, satisfaction levels, and product preferences could provide deeper insights into optimizing marketing efforts and improving customer lifetime value.
By leveraging these insights, insurance companies can refine their strategies to effectively cater to different customer segments, thereby maximizing customer lifetime value and overall business performance.

## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [ ]:
marketing_df.value_counts()


unnamed:_0  customer  state       customer_lifetime_value  response  coverage  education             effective_to_date  employment_status  gender  income  location_code  marital_status  monthly_premium_auto  months_since_last_claim  months_since_policy_inception  number_of_open_complaints  number_of_policies  policy_type     policy        renew_offer_type  sales_channel  total_claim_amount  vehicle_class  vehicle_size  vehicle_type  month
0           DK49336   Arizona     4809.216960              No        Basic     College               2011-02-18         Employed           M       48029   Suburban       Married         61                    7.0                      52                             0.0                        9                   Corporate Auto  Corporate L3  Offer3            Agent          292.800000          Four-Door Car  Medsize       A             2        1
7276        WP55297   Washington  4629.970602              No        Basic     Bachelor              2011-01-

In [ ]:
marketing_df.pivot_table(values='number_of_open_complaints', columns= 'policy_type', index=['month'], aggfunc='count')



In [ ]:
#stack the pivot table to create a long format table
marketing_df.pivot_table.stack().reset_index()